server/sonar-server/src/main/java/org/sonar/server/qualityprofile/QProfileResetImpl.java

```diff
@@ -0,0 +1,158 @@
+  }
+
+  @Override
+  public void resetLanguage(DbSession dbSession, OrganizationDto organization, String language) {
+    ListMultimap<QProfileName, RulesProfile> profilesByName = loadDefinitionsGroupedByName(language);
+    for (Map.Entry<QProfileName, Collection<RulesProfile>> entry : profilesByName.asMap().entrySet()) {
+      QProfileName profileName = entry.getKey();
+      QualityProfileDto profile = factory.getOrCreate(dbSession, organization, profileName);
+      List<RuleActivation> activations = Lists.newArrayList();
+      for (RulesProfile def : entry.getValue()) {
+        for (ActiveRule activeRule : def.getActiveRules()) {
+          activations.add(getRuleActivation(dbSession, activeRule));
+        }
+      }
+      reset(dbSession, profile, activations);
+    }
+  }
+
+  private RuleActivation getRuleActivation(DbSession dbSession, ActiveRule activeRule) {
+    RuleActivation activation = new RuleActivation(RuleKey.of(activeRule.getRepositoryKey(), activeRule.getRuleKey()));
+    activation.setSeverity(activeRule.getSeverity().name());
+    if (!activeRule.getActiveRuleParams().isEmpty()) {
+      for (ActiveRuleParam param : activeRule.getActiveRuleParams()) {
+        activation.setParameter(param.getParamKey(), param.getValue());
+      }
+    } else {
+      for (RuleParamDto param : db.ruleDao().selectRuleParamsByRuleKey(dbSession, activeRule.getRule().ruleKey())) {
+        activation.setParameter(param.getName(), param.getDefaultValue());
+      }
+    }
+    return activation;
+  }
+
+  @Override
+  public BulkChangeResult reset(DbSession dbSession, QualityProfileDto profile, Collection<RuleActivation> activations) {
+    requireNonNull(profile.getId(), "Quality profile must be persisted");
+    BulkChangeResult result = new BulkChangeResult();
+    Set<RuleKey> ruleToBeDeactivated = Sets.newHashSet();
+    // Keep reference to all the activated rules before backup restore
+    for (ActiveRuleDto activeRuleDto : db.activeRuleDao().selectByProfileKey(dbSession, profile.getKee())) {
+      if (activeRuleDto.getInheritance() == null) {
+        // inherited rules can't be deactivated
+        ruleToBeDeactivated.add(activeRuleDto.getKey().ruleKey());
+      }
+    }
+
+    for (RuleActivation activation : activations) {
+      try {
+        List<ActiveRuleChange> changes = activator.activate(dbSession, activation, profile.getKey());
+        ruleToBeDeactivated.remove(activation.getRuleKey());
+        result.incrementSucceeded();
+        result.addChanges(changes);
+      } catch (BadRequestException e) {
+        result.incrementFailed();
+        result.getErrors().addAll(e.errors());
+      }
+    }
+
+    List<ActiveRuleChange> changes = new ArrayList<>();
+    changes.addAll(result.getChanges());
+    for (RuleKey ruleKey : ruleToBeDeactivated) {
+      try {
+        changes.addAll(activator.deactivate(dbSession, ActiveRuleKey.of(profile.getKee(), ruleKey)));
+      } catch (BadRequestException e) {
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png 'Severity: MAJOR') Either log or rethrow this exception. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS1166)